src/main/java/com/google/sps/utility/FreeTimeUtility.java

```diff
@@ -0,0 +1,142 @@
+    this.eveningFreeInterval = updateInterval(this.eveningFreeInterval, eventStart, eventEnd);
+  }
+
+  /**
+   * Method to update the free intervals in a particular list of intervals.
+   *
+   * @param freeTime the list of intervals to update.
+   * @param eventStart the start time of the new event
+   * @param eventEnd the end time of the new event
+   * @return the new list of intervals
+   */
+  private List<DatePair> updateInterval(List<DatePair> freeTime, Date eventStart, Date eventEnd) {
+    List<DatePair> updatedTime = new ArrayList<>();
+    for (DatePair interval : freeTime) {
+      if (interval.getKey().before(eventStart) && interval.getValue().after(eventEnd)) {
+        updatedTime.add(new DatePair(interval.getKey(), eventStart));
+        updatedTime.add(new DatePair(eventEnd, interval.getValue()));
+      } else if (interval.getKey().before(eventStart) && interval.getValue().after(eventStart)) {
+        updatedTime.add(new DatePair(interval.getKey(), eventStart));
+      } else if (interval.getKey().before(eventEnd) && interval.getValue().after(eventEnd)) {
+        updatedTime.add(new DatePair(eventEnd, interval.getValue()));
+      } else if (!(eventStart.before(interval.getKey()) && eventEnd.after(interval.getValue()))) {
+        updatedTime.add(interval);
+      }
+    }
+    return updatedTime;
+  }
+
+  /** Method to calculate the free time for each day based on the free intervals. */
+  private void updateFreeTime() {
+    long zero = 0;
+    this.workHoursPerDay = Arrays.asList(zero, zero, zero, zero, zero);
+    this.personalHoursPerDay = Arrays.asList(zero, zero, zero, zero, zero);
+    this.workHoursPerDay = updateHoursPerDay(this.workHoursPerDay, this.workFreeInterval);
+    this.personalHoursPerDay =
+        updateHoursPerDay(this.personalHoursPerDay, this.morningFreeInterval);
+    this.personalHoursPerDay =
+        updateHoursPerDay(this.personalHoursPerDay, this.eveningFreeInterval);
+  }
+
+  /**
+   * Method to update the free time on a particular list of free time based on a particular list of
+   * intervals
+   *
+   * @param hoursPerDay the list of free time in milliseconds to update.
+   * @param freeInterval the list of free intervals
+   * @return the updated list of free time
+   */
+  private List<Long> updateHoursPerDay(List<Long> hoursPerDay, List<DatePair> freeInterval) {
+    for (DatePair interval : freeInterval) {
+      int index = interval.getKey().getDay() - this.startDate.getDay();
+      index = index < 0 ? index + OFFSET : index;
+      hoursPerDay.set(
+          index,
+          hoursPerDay.get(index) + interval.getValue().getTime() - interval.getKey().getTime());
+    }
+    return hoursPerDay;
+  }
+
+  /**
+   * Method to create and return the calendar data response.
+   *
+   * @return the calendar data response to be sent as servlet response
+   */
+  public CalendarDataResponse getCalendarDataResponse() {
+    updateFreeTime();
```

Comment:
I think it would be cleaner to inline the `updateFreeTime` logic here. It's only used in this one place, and it's what produces the data you return, so having it here makes sense.

Reply:
I agree. `Resolved`